Data/ByteString/Internal.hs

```diff
@@ -339,18 +339,18 @@ packChars cs = unsafePackLenChars (List.length cs) cs
  #-}
 
 unsafePackLenBytes :: Int -> [Word8] -> ByteString
-unsafePackLenBytes len xs0 =
-    unsafeCreate len $ \p -> go p xs0
-  where
-    go !_ []     = return ()
-    go !p (x:xs) = poke p x >> go (p `plusPtr` 1) xs
+unsafePackLenBytes len =
+    unsafeCreate len . foldr
+      (\x go p -> poke p x >> go (p `plusPtr` 1))
+      (\_ -> return ())
+{-# INLINE unsafePackLenBytes #-}
 
 unsafePackLenChars :: Int -> [Char] -> ByteString
-unsafePackLenChars len cs0 =
-    unsafeCreate len $ \p -> go p cs0
-  where
-    go !_ []     = return ()
-    go !p (c:cs) = poke p (c2w c) >> go (p `plusPtr` 1) cs
+unsafePackLenChars len =
+    unsafeCreate len . foldr
+      (\x go p -> poke p (c2w x) >> go (p `plusPtr` 1))
+      (\_ -> return ())
+{-# INLINE unsafePackLenChars #-}
```

Comment:
Does the inline pragma cause this code to be inlined into `ByteString` literals. If so, what are the compile time implications? Compare https://github.com/jgm/doclayout/pull/1.

The inline pragma for `unsafePackLenBytes` may have similar effects for `Bytestring` list literals?

Reply:
There's a rewrite rule that must fire before unsafePackLenBytes can be inlined:

https://github.com/haskell/bytestring/blob/12563781f257f132f4da866047abe720623f9990/Data/ByteString/Internal.hs#L336-L339